Sources/Sentry/SentryTracer.m

```diff
@@ -7,25 +7,75 @@
 #import "SentryTransaction.h"
 #import "SentryTransactionContext.h"
 
+@interface
+SentryTracer ()
+
+/**
+ * Perform a check whether this trace can be finished, if so, finishes the trace.
+ *
+ * The tracer can be finished when _waitChildren is NO or
+ * all children are finished and the finish function was called at least once.
+ *
+ */
+- (void)checkFinish;
+
+/**
+ * Returns a flat list of all children recursively.
+ */
+- (NSArray *)children;
+
+/**
+ * A lock to coordinate child manipulation.
+ */
+- (NSLock *)childrenLock;
+@end
+
 @implementation SentryTracer {
     SentrySpan *_rootSpan;
     NSMutableArray<id<SentrySpan>> *_spans;
     SentryHub *_hub;
+    SentrySpanStatus _finishStatus;
+    BOOL _canBeFinished;
+    BOOL _waitChildren;
+    SentryTracer *_parentTracer;
+    NSLock *_childrenLock;
```

Comment:
`m`: This whole class is not thread-safe, why do we need locking now?

Reply:
Well, let's make it thread safe because it will definitely need to be with this new logic.

Reply:
I did find a potencial problem in SentrySpan and SentrySpanContext, but not in SentryTracer regarding thread safety, could you point it out what am I missing?